src/main/java/com/winner_is_kungen/tda367/model/Component.java

```diff
@@ -91,20 +84,19 @@ public int getNrOutputs() {
 	 * @param outChannel A Integer specifying which input is used.
 	 */
 	void addListener(ComponentListener listener, int inChannel, int outChannel) {
-		listeners.add(new Tuple<>(listener, inChannel, outChannel));
-
-		listener.update(logic(inputChannels)[outChannel], inChannel);
+		signal.add(new Tuple<>(listener, inChannel, outChannel));
+		signal.broadcastUpdate(UUID.randomUUID().toString(),logic(inputChannels));
```

Comment:
This updates all listeners, right?

Do we want that? **I** don't think so. But I also don't think the old way was better. Maybe the `Signal` class could cache all channels and when a listener is added call that listener with the cached value.

I think that recalculating (the `logic` call) because of a new output would be bad. I think we only should recalculate when the component receives new input.

Reply:
This is just a comment. I want somebody else's  input too.

Reply:
The question is what takes more, recalculating the logic call or saving the outputs of every logic call.

 I agree that calling update on all the listeners is a waste of time, but i think caching the result of all logic calls is going to make a bigger performance hit than just rerunning logic. Especially since we will be calling logic with new values more often than not.

Either way ill change it so that is only updates the relevant listener.

Reply:
It's not a question of performance. If we have a self-calling configuration the output from a component could change because of a new output connection. Which is a behavior i don't think we want.

Reply:
This sound more like a question of whether the logic function should be allowed to have a state. If it doesn't then there would be no difference between calling logic with the cached input and just caching the output.

If they are allowed to have a state then the need for an output cache seems like a reasonable solution.

In my personal opinion, making logic stateless and have the current state in the input cache seems cleaner. This would require us to make any component with states out of the primary components ( And, Or, Not etc) as custom components, but i like the idea that every complicated component could be easily broken down into its core logical components.